### ISSUE TITLE
Support Integer/Float Min/Max Operations

### ISSUE THREAD

**Issue by naftulikay** [Keats/tera#433]
A feature that would be super helpful would be to implement a `min` and `max` function for generic number types.

First stage:

 - unsigned integers: upgrade all unsigned integers to a `u64` or a `usize`
 - floats: upgrade all floats to a `f64`
 - non-numbers: return an `Err`

Second stage:

 - unsigned/signed integer comparison: pretty straightforward comparison
 - float/integer comparison: also fairly straightforward

This would probably be a useful addition to the 1.0.0 release. I'd like to take on the work, but my availability isn't great right now.

**Comment by Keats** [comment]
That does seem a bit niche and is easy to implement in someone's project if they need it. I don't think it makes sense to add it to Tera itself

**Comment by naftulikay** [comment]
I know that Tera is not Jinja, [but Jinja does support these operations](http://jinja.pocoo.org/docs/2.10/templates/#max).

The specific use-case that we had was rendering two different systemd unit drop-ins to isolate two processes by CPU:

Template 1:

```jinja
[Service]
CPUAffinity=0-{{ max(0, floor(num_cpus / 2) - 1) }} 
```

Template 2:

```jinja
[Service]
CPUAffinity={{ max(0, floor(num_cpus / 2)) }}-{{ max(0, num_cpus - 1) }}
```

There are edge-cases which this template does not catch, but the idea would be that service A would be pinned from CPUs 0-7 and service B to CPUs 8-15 on a 16-core machine.

I'm still happy to contribute `min`, `max`, and `floor` as either filters or functions to Tera to add this functionality if there's a desire for it.

**Comment by Keats** [comment]
hmm If they exist in Jinja2 it's different. The main issue is that we can't really support variadic functions so we are limited to 2 arguments and a function wouldn't be able to get the min/max of an array.

**Comment by naftulikay** [comment]
Yes, my usage above was pseudocode, it's likely that these functions would need to be filters like Jinja does it.

**Comment by Keats** [comment]
Closing as part of the Tera v2 release. If this is still relevant in v2, please reopen — see the 
  migration guide: https://github.com/Keats/tera/blob/master/MIGRATION.md
